### PR TITLE
Don’t hash datasets

### LINKED ISSUE #4146: Caching a Dataset for Multiple Specs
I have a use case that a user can create different visualizations with a single dataset in a trial-and-error fashion. I tried inlining the parsed dataset into the specs but it didn't seem to help. From the Chrome profiling tool, Vega-lite always "parse" the dataset even if it's inline javascript objects. Relevant screenshots are provided below.

This report is created according to the stackoverflow thread: [https://stackoverflow.com/questions/51455030/cache-a-dataset-in-vega-or-vega-embed](https://stackoverflow.com/questions/51455030/cache-a-dataset-in-vega-or-vega-embed)

The spec as shown by `console.log`, which shows the data is indeed inline.
vega_lite_spec_inline_data

The screenshot of Chrome's profiling tool shows that the `parse` function takes most of the time.
vega_lite_stack

As this is my first time creating a bug report, I hope I am not breaking any rules. Let me know if I can provide more materials.

### PR DESCRIPTION
**Pull request by domoritz**
Fixes #4146.

### CODE REVIEW

File: src/compile/data/index.ts [hunk 1]
```diff
@@ -7,9 +7,9 @@ import {SourceNode} from './source';
 
 export interface DataComponent {
   /**
-   * A dictionary of sources indexed by a hash.
+   * A list on unique sources.
```
Comment (invokesus) [inline src/compile/data/index.ts]: *of

File: src/compile/data/parse.ts [hunk 2]
```diff
@@ -39,17 +39,41 @@ import {TimeUnitNode} from './timeunit';
...
+export function findSource(data: Data, sources: SourceNode[]) {
+  for (const other of sources) {
+    const otherData = other.data;
+    if (isInlineData(data) && isInlineData(otherData)) {
+      const srcVals = data.values;
+      const otherVals = otherData.values;
+      if (deepEqual(srcVals, otherVals)) {
+        return other;
+      }
+    } else if (isUrlData(data) && isUrlData(otherData)) {
```
Comment (invokesus) [inline src/compile/data/parse.ts]: How about `else if (isUrlData(data) && isUrlData(otherData) && data.url === otherData.url) {` ?

File: src/compile/data/optimize.ts [hunk 3]
```diff
@@ -159,8 +159,9 @@ export function isTrue(x: boolean) {
 }
 
 function optimizationDataflowHelper(dataComponent: DataComponent) {
-  let roots: SourceNode[] = vals(dataComponent.sources);
+  let roots: SourceNode[] = dataComponent.sources;
```
Comment (invokesus) [inline src/compile/data/optimize.ts]: This is a copy by reference.

File: src/compile/data/optimize.ts [hunk 4]
```diff
@@ -185,11 +186,9 @@ function optimizationDataflowHelper(dataComponent: DataComponent) {
 
   mutatedFlag = roots.map(mergeParse).some(isTrue) || mutatedFlag;
   mutatedFlag = roots.map(optimizers.mergeIdenticalNodes).some(isTrue) || mutatedFlag;
-  keys(dataComponent.sources).forEach(s => {
-    if (dataComponent.sources[s].numChildren() === 0) {
-      delete dataComponent.sources[s];
-    }
-  });
+
+  dataComponent.sources = roots;
```
Comment (invokesus) [inline src/compile/data/optimize.ts]: Both are the same array.

File: src/compile/data/parse.ts (hunk 2 above)
Comment (invokesus) [reply src/compile/data/parse.ts]: Oh, the type narrowing won't work with this I guess.

File: src/compile/data/optimize.ts (hunk 4 above)
Comment (domoritz) [reply src/compile/data/optimize.ts]: No, we use filter to remove things from the array and that is not in place.

File: src/compile/data/optimize.ts (hunk 3 above)
Comment (domoritz) [reply src/compile/data/optimize.ts]: See comment below

**Comment by domoritz** [comment]
Rebased

File: src/compile/data/index.ts (hunk 1 above)
Comment (domoritz) [reply src/compile/data/index.ts]: Fixed

**Pull request merged**